src/componentSetBuilder.ts

```diff
@@ -38,64 +37,69 @@ export class ComponentSetBuilder {
       if (packagenames) {
         logger.debug(`Building ComponentSet for packagenames: ${packagenames.toString()}`);
-        csAggregator.push(...new ComponentSet([]));
+        componentSet ??= new ComponentSet();
       }
 
       // Resolve manifest with source in package directories.
       if (manifest) {
         logger.debug(`Building ComponentSet from manifest: ${manifest.manifestPath}`);
         const directoryPaths = options.manifest.directoryPaths;
         logger.debug(`Searching in packageDir: ${directoryPaths.join(', ')} for matching metadata`);
-        const compSet = await ComponentSet.fromManifest({
+        componentSet = await ComponentSet.fromManifest({
           manifestPath: manifest.manifestPath,
           resolveSourcePaths: options.manifest.directoryPaths,
           forceAddWildcards: true,
         });
-        csAggregator.push(...compSet);
       }
 
       // Resolve metadata entries with source in package directories.
       if (metadata) {
         logger.debug(`Building ComponentSet from metadata: ${metadata.metadataEntries.toString()}`);
         const registry = new RegistryAccess();
+        const compSetFilter = new ComponentSet();
+        componentSet ??= new ComponentSet();
 
         // Build a Set of metadata entries
-        const filter = new ComponentSet();
-        metadata.metadataEntries.forEach((entry) => {
-          const splitEntry = entry.split(':');
-          // try and get the type by name to ensure no typos or errors in type name
-          // matches toolbelt functionality
+        metadata.metadataEntries.forEach((rawEntry) => {
+          const splitEntry = rawEntry.split(':');
+          // The registry will throw if it doesn't know what this type is.
           registry.getTypeByName(splitEntry[0]);
-          filter.add({
+          const entry = {
             type: splitEntry[0],
             fullName: splitEntry.length === 1 ? '*' : splitEntry[1],
-          });
+          };
+          // Add to the filtered ComponentSet for resolved source paths,
+          // and the unfiltered ComponentSet to build the correct manifest.
+          compSetFilter.add(entry);
+          componentSet.add(entry);
         });
 
         const directoryPaths = options.metadata.directoryPaths;
         logger.debug(`Searching for matching metadata in directories: ${directoryPaths.join(', ')}`);
-        const fromSource = ComponentSet.fromSource({ fsPaths: directoryPaths, include: filter });
-        // If no matching metadata is found, default to the original component set
-        const finalized = fromSource.size > 0 ? fromSource : filter;
-        csAggregator.push(...finalized);
+        const resolvedComponents = ComponentSet.fromSource({ fsPaths: directoryPaths, include: compSetFilter });
+        componentSet.forceIgnoredPaths = resolvedComponents.forceIgnoredPaths;
+        for (const comp of resolvedComponents) {
```

Comment:
ooh, I didn't know you could iterate it without turning it into an array or calling some other method.  Nice.

Reply:
I didn't either until I saw it done that way in SDR.